lib/SILOptimizer/Transforms/AccessEnforcementOpts.cpp

```diff
@@ -401,146 +575,231 @@ void AccessConflictAnalysis::identifyBeginAccesses() {
-    if (!callSiteAccesses.mayConflictWith(accessKind, outerAccess))
-      continue;
-
-    LLVM_DEBUG(llvm::dbgs() << *fullApply.getInstruction()
-                            << "  call site access: ";
-               callSiteAccesses.dump();
-               llvm::dbgs() << "  may conflict with:\n";
-               outerAccess.dump());
-
-    recordConflict(outerAccess, accessSet);
-  }
+  detectApplyConflicts(applyStorage->getInScopeAccesses(), applyStorage,
+                       callSiteAccesses, fullApply);
+  detectApplyConflicts(applyStorage->getOutOfScopeConflictFreeAccesses(),
+                       applyStorage, callSiteAccesses, fullApply);
 }
 
-// Merge all predecessor accesses into the local acces set. Only propagate
-// accesses that are still present in all predecessors. The absence of a begin
-// access from a visited predecessor indicates the presence of a conflict. A
-// block has been visited if it has a map entry in blockOutAccess.
-void AccessConflictAnalysis::mergePredAccesses(SILBasicBlock *succBB,
-                                      SparseAccessSet &mergedAccesses) {
-  for (SILBasicBlock *predBB : succBB->getPredecessorBlocks()) {
-    auto mapI = blockOutAccess.find(predBB);
-    if (mapI == blockOutAccess.end())
-      continue;
+void RegionStorageAnalysis::visitEndAccess(EndAccessInst *endAccess) {
+  auto *beginAccess = endAccess->getBeginAccess();
+  if (beginAccess->getEnforcement() != SILAccessEnforcement::Dynamic)
+    return;
 
-    const DenseAccessSet &predSet = mapI->second;
-    mergedAccesses.merge(SparseAccessSet(predSet, result));
-  }
-}
+  auto *endStorage = getRegionStorageFromIns(endAccess);
+  auto &inScope = endStorage->getInScopeAccesses();
+  if (std::find(inScope.begin(), inScope.end(), beginAccess) == inScope.end())
+    return;
 
-// Compute access reachability within the given block.
-void AccessConflictAnalysis::visitBlock(SILBasicBlock *BB) {
-  // Sparse set for tracking accesses with an individual block.
-  SparseAccessSet accessSet(result);
-  mergePredAccesses(BB, accessSet);
+  LLVM_DEBUG(llvm::dbgs() << "No conflict on one path from " << *beginAccess
+                          << " to " << *endAccess);
+  endStorage->removeInScopeAccess(beginAccess);
+  LLVM_DEBUG(llvm::dbgs() << "Got out of scope from " << *beginAccess << " to "
+                          << *endAccess
+                          << " with no conflicts seen (so far)\n");
+  endStorage->addOutOfScopeAccess(beginAccess);
+}
 
-  for (auto &I : *BB) {
+void RegionStorageAnalysis::processBlockRegion(unsigned id) {
+  assert(regionToStorageMap.find(id) == regionToStorageMap.end() &&
+         "Should not process a region twice");
+  auto *region = LRFI->getRegion(id);
+  assert(region->isBlock() && "Expected a block");
+  // A block's inital set is all accesses
+  auto *storage = new RegionStorage(region, result, storageAccesses);
+  regionToStorageMap.insert(std::make_pair(id, storage));
```

Comment:
You can just assert that the result value (.second) is true rather than doing the extra `find(id)` above.